charms/kfp-viz/tests/unit/test_operator.py

```diff
@@ -15,83 +14,117 @@
 
 
-def test_image_fetch(harness, oci_resource_data):
+def test_pebble_service_container_running(harness, mocked_kubernetes_service_patch):
+    """Test that the pebble service of the charm's kfp-visualization container is running."""
+    # Arrange
+    harness.set_leader(True)
     harness.begin()
-    with pytest.raises(MissingResourceError):
-        harness.charm.image.fetch()
-
-    harness.add_oci_resource(**oci_resource_data)
-    with does_not_raise():
-        harness.charm.image.fetch()
+    harness.set_can_connect("ml-pipeline-visualizationserver", True)
 
+    # Mock:
+    # * leadership_gate to have get_status=>Active
+    harness.charm.leadership_gate.get_status = MagicMock(return_value=ActiveStatus())
 
-# TODO: kfp-viz (provide data)
+    # Act
+    harness.charm.on.install.emit()
 
+    assert isinstance(harness.charm.unit.status, ActiveStatus)
 
-def test_install_with_all_inputs(harness, oci_resource_data):
-    harness.set_leader()
-    http_port = "1234"
-    model_name = "test_model"
-    harness.set_model_name(model_name)
-    harness.update_config({"http-port": http_port})
+    # Assert
+    container = harness.charm.unit.get_container("ml-pipeline-visualizationserver")
+    # Assert that sidecar container is up and its service is running
+    assert container.get_service("vis-server").is_running()
 
-    kfpviz_relation_name = "kfp-viz"
 
-    harness.set_leader()
+def test_kfp_viz_relation_with_related_app(harness, mocked_kubernetes_service_patch):
+    """Test that the kfp-viz relation sends data to related apps and goes Active."""
+    # Arrange
+    harness.set_leader(True)  # needed to write to an SDI relation
+    model = "model"
+    harness.set_model_name(model)
+    harness.begin()
 
-    # Set up required relations
-    # Future: convert these to fixtures and share with the tests above
-    harness.add_oci_resource(**oci_resource_data)
+    # Mock:
+    # * leadership_gate to be active and executed
+    harness.charm.leadership_gate.get_status = MagicMock(return_value=ActiveStatus())
+
+    expected_relation_data = {
+        "_supported_versions": ["v1"],
+        "data": render_kfp_viz_data(
+            app_name=harness.model.app.name,
+            model_name=model,
+            port=harness.model.config["http-port"],
+        ),
+    }
 
-    relation_version_data = {"_supported_versions": "- v1"}
+    # Act
+    # Add one relation with data.  This should trigger a charm reconciliation due to
+    # relation-changed.
+    relation_metadata = add_sdi_relation_to_harness(harness, "kfp-viz", other_app="o1", data={})
+    relation_ids_to_assert = [relation_metadata.rel_id]
+
+    # Assert
+    assert isinstance(harness.charm.kfp_viz_relation.status, ActiveStatus)
+    assert_relation_data_send_as_expected(harness, expected_relation_data, relation_ids_to_assert)
+
+
+def assert_relation_data_send_as_expected(harness, expected_relation_data, rel_ids_to_assert):
```

Comment:
Just because `harness` is a fixture, you want to make a clear distinction between the fixture and an argument.
 
```suggestion
def assert_relation_data_send_as_expected(harness_object, expected_relation_data, rel_ids_to_assert):
```